BlueprintUICommonControls/Sources/AttributedLabel.swift

```diff
@@ -43,53 +57,502 @@ public struct AttributedLabel: Element, Hashable {
+            case .link: return .link
+            case .phoneNumber: return .phoneNumber
             }
-            label.accessibilityTraits = union
         }
     }
 }
 
+// MARK: View implementation
+
 extension AttributedLabel {
 
     private final class LabelView: UILabel {
-        var textRectOffset: UIOffset = .zero {
+        /// The touch handling logic explicitly tracks the active links when touches begin, so if you drag outside
+        /// the link and touch up over another link, it just cancels the tap rather than accidentally opening
+        /// a different link.
+        private var trackingLinks: [Link]?
+
+        private var linkDetectionTypes: Set<LinkDetectionType> = []
+        private var linkAttributes: [NSAttributedString.Key: Any] = [:]
+        private var activeLinkAttributes: [NSAttributedString.Key: Any] = [:]
+
+        private var links: [Link] = [] {
+            didSet {
+                isUserInteractionEnabled = !links.isEmpty
+            }
+        }
+
+        private var textRectOffset: UIOffset = .zero {
             didSet {
                 if oldValue != textRectOffset {
                     setNeedsDisplay()
                 }
             }
         }
 
+        /// The view needs to keep track of the current index for the accessibility rotor.
+        private var accessibilityLinkIndex = -1
+
+        /// These elements need to be retained by the view, and cannot be created inside the
+        /// `accessibilityCustomRotors` getter.
+        private var accessibilityLinks: [LinkAccessibilityElement] = []
+
+        override var accessibilityCustomRotors: [UIAccessibilityCustomRotor]? {
+            set { fatalError() }
+            get {
+                [
+                    UIAccessibilityCustomRotor(systemType: .link) { [weak self] predicate in
+                        guard let self = self else {
+                            return nil
+                        }
+
+                        self.accessibilityLinkIndex += predicate.searchDirection == .next ? 1 : -1
+                        self.accessibilityLinkIndex = min(self.accessibilityLinks.count - 1, self.accessibilityLinkIndex)
+                        self.accessibilityLinkIndex = max(0, self.accessibilityLinkIndex)
+
+                        let link = self.accessibilityLinks[self.accessibilityLinkIndex]
+                        return UIAccessibilityCustomRotorItemResult(targetElement: link, targetRange: nil)
+                    },
+                ]
+            }
+        }
+
+        var linkHandler: LinkHandler?
+
+        func update(model: AttributedLabel, linkHandler: LinkHandler?, isMeasuring: Bool) {
+            let previousString = attributedText?.string
+
+            linkAttributes = model.linkAttributes
+            activeLinkAttributes = model.activeLinkAttributes
+            linkDetectionTypes = model.linkDetectionTypes ?? []
+            attributedText = model.attributedText
+                .applyingDefaultFont()
```

Comment:
Curious about this — are we relying on the presence of font attributes?

Reply:
The attributed string needs a font attribute specified across the whole thing in order for the layout manager to match the label - I initially was just setting the attribute of the attributed string when I measure it, and setting it to the `font` property, but the rect was off with that approach (not really sure why, but I think the default `font` is special in some way). Explicitly using a default font for the label like this fixes that issue (and matches the default font of UILabel appearance-wise).

Reply:
Oh, actually, I can tell you why. NSAttributedString has a [defined default font of 12-point Helvetica](https://developer.apple.com/documentation/foundation/nsattributedstring/key/1528839-font).

> If you do not specify this attribute, the string uses a 12-point Helvetica(Neue) font by default.

And UILabel has its own default, probably because it predates attributed strings.

Reply:
Right, I didn't expect the defaults to match; but adding `self.font` as the `font` attribute when before feeding the string to layout manager wasn't working as expected, which is the part I didn't really get. It's possible I was doing something else wrong (this was early in my explorations) but this seems like a good "safe" way to make sure they match